cedar-lean/Cedar/Thm/Partial/EvaluateCall.lean

```diff
@@ -0,0 +1,178 @@
+import Cedar.Thm.Data.Set
+import Cedar.Thm.Partial.Evaluation.Props
+import Cedar.Thm.Partial.Evaluation.Set
+import Cedar.Thm.Partial.WellFormed
+import Cedar.Thm.Partial.Subst
+
+/-! Theorems about `Partial.evaluateCall` -/
+
+namespace Cedar.Thm.Partial.EvaluateCall
+
+open Cedar.Data
+open Cedar.Partial (Subsmap Unknown)
+open Cedar.Spec (Error Expr Ext ExtFun Prim Result)
+
+/--
+  `Partial.evaluateCall` on concrete arguments gives the same output as
+  `Spec.call` on those same arguments
+-/
+theorem on_concrete_eqv_concrete {vs : List Spec.Value} {xfn : ExtFun} :
+  Partial.evaluateCall xfn (vs.map Partial.Value.value) = (Spec.call xfn vs).map Partial.Value.value
+:= by
+  unfold Partial.evaluateCall
+  simp only [List.mapM_map, List.mapM_some, Except.map]
+  cases Spec.call xfn vs <;> simp
+
+/--
+  if `Spec.call` returns `ok` with some value, that is a well-formed value
+-/
+theorem specCall_wf {vs : List Spec.Value} {xfn : ExtFun}
+  (wf : ∀ v ∈ vs, v.WellFormed) :
+  ∀ v, Spec.call xfn vs = .ok v → v.WellFormed
+:= by
+  unfold Spec.Value.WellFormed
+  intro v
+  cases v <;> simp
+  case prim p => simp [Prim.WellFormed]
+  case set | record =>
+    unfold Spec.call Spec.res
+    split <;> simp at * <;> split <;> simp
+  case ext x => cases x <;> simp [Ext.WellFormed]
+
+/--
+  if `Partial.evaluateCall` on well-formed arguments returns `ok` with some
+  value, that is a well-formed value as well
+-/
+theorem evaluateCall_wf {pvals : List Partial.Value} {xfn : ExtFun}
+  (wf : ∀ pval ∈ pvals, pval.WellFormed) :
+  ∀ pval, Partial.evaluateCall xfn pvals = .ok pval → pval.WellFormed
+:= by
+  unfold Partial.evaluateCall Partial.Value.WellFormed Partial.ResidualExpr.WellFormed
+  intro pval h₁
+  split at h₁
+  · rename_i vs h₂
+    cases h₃ : Spec.call xfn vs <;> simp [h₃] at h₁
+    subst pval
+    rename_i v'
+    apply specCall_wf _ v' h₃
+    intro v h₅
+    replace ⟨pval, h₄, h₂⟩ := List.mapM_some_implies_all_from_some h₂ v h₅
+    specialize wf pval h₄
+    unfold Partial.Value.WellFormed at wf
+    cases pval <;> simp at wf h₂
+    case value v' => subst v' ; exact wf
+  · simp only [Except.ok.injEq] at h₁ ; subst h₁ ; simp only
+
+/--
+  If `Partial.evaluateCall` produces `ok` with a concrete value, then all of the
+  arguments are concrete
+-/
+theorem returns_concrete_then_args_concrete {args : List Partial.Value} {xfn : ExtFun} :
+  Partial.evaluateCall xfn args = .ok (.value v) →
+  ∀ arg ∈ args, ∃ v, arg = .value v
+:= by
+  unfold Partial.evaluateCall
+  split <;> intro h₁ arg h₂
+  · rename_i vs h₃
+    replace ⟨v, h₃, h₄⟩ := List.mapM_some_implies_all_some h₃ arg h₂
+    cases arg <;> simp only [Option.some.injEq] at h₄
+    subst v ; rename_i v
+    exists v
+  · rename_i h₃
+    replace ⟨arg', h₃, h₄⟩ := List.mapM_none_iff_exists_none.mp h₃
+    cases arg' <;> simp only [Except.ok.injEq] at h₁ h₄
+
+/--
+  something akin to `Partial.EvaluateValue.eval_spec_value`, lifted to lists of `Partial.Value`
```

Comment:
I don't think there's any `EvaluateValue` file in the repo currently. Coming in a future commit?

Reply:
Yep.  Right now the comment should be read as "if we were to have such a thing".

Alternately, I can add the file in this PR with just that theorem (which has a 1-line proof) even though it isn't needed by any of the proofs that exist today on `main`

Reply:
I think it'd be better to wait to add this lemma until you add the proof that requires it. But fine either way 🙂